src/Contracts/OpdexVault.cs

```diff
@@ -81,129 +136,348 @@ public void NotifyDistribution(UInt256 amount)
         Assert(Message.Sender == Token, "OPDEX: UNAUTHORIZED");
 
         TotalSupply += amount;
-
-        if (Genesis == 0) Genesis = Block.Number;
     }
 
     /// <inheritdoc />
-    public void CreateCertificate(Address to, UInt256 amount)
+    public ulong CreateNewCertificateProposal(UInt256 amount, Address recipient, string description)
     {
-        var owner = Owner;
-        var vestingDuration = VestingDuration;
+        ValidateProposal(amount, description);
 
-        Assert(Message.Sender == owner, "OPDEX: UNAUTHORIZED");
-        Assert(to != owner, "OPDEX: INVALID_CERTIFICATE_HOLDER");
-        Assert(amount > 0 && amount <= TotalSupply, "OPDEX: INVALID_AMOUNT");
-        Assert(Block.Number < Genesis + vestingDuration, "OPDEX: TOKENS_BURNED");
+        var certificate = GetCertificate(recipient);
 
-        var certificates = GetCertificates(to);
+        Assert(certificate.VestedBlock == 0ul, "OPDEX: CERTIFICATE_EXISTS");
 
-        Assert(certificates.Length < MaximumCertificates, "OPDEX: CERTIFICATE_LIMIT_REACHED");
+        var totalProposedAmount = TotalProposedAmount;
 
-        var vestedBlock = Block.Number + vestingDuration;
+        Assert(TotalSupply - totalProposedAmount >= amount, "OPDEX: INSUFFICIENT_VAULT_SUPPLY");
 
-        certificates = InsertCertificate(certificates, amount, vestedBlock, false);
+        TotalProposedAmount = totalProposedAmount + amount;
 
-        SetCertificates(to, certificates);
+        return CreateCertificateTypeProposalExecute(recipient, amount, description, (byte)ProposalType.Create);
+    }
 
-        TotalSupply -= amount;
+    /// <inheritdoc />
+    public ulong CreateRevokeCertificateProposal(Address recipient, string description)
+    {
+        var certificate = GetCertificate(recipient);
+        var maxExpiration = Block.Number + VoteDuration + PledgeDuration;
 
-        Log(new CreateVaultCertificateLog{ Owner = to, Amount = amount, VestedBlock = vestedBlock });
+        Assert(!certificate.Revoked && certificate.VestedBlock > maxExpiration, "OPDEX: INVALID_CERTIFICATE");
+        ValidateProposal(certificate.Amount, description);
+        return CreateCertificateTypeProposalExecute(recipient, certificate.Amount, description, (byte)ProposalType.Revoke);
     }
 
     /// <inheritdoc />
-    public void RedeemCertificates()
+    public ulong CreatePledgeMinimumProposal(UInt256 amount, string description)
     {
-        var certificates = GetCertificates(Message.Sender);
-        var lockedCertificates = new VaultCertificate[0];
-        var amountToTransfer = UInt256.Zero;
+        ValidateProposal(amount, description);
```

Comment:
I believe we should also validate that the proposal isn't proposing to change the pledge/vote minimum to the same amount. Otherwise you're needlessly creation a proposal.

Reply:
An extra check that won't hurt anything except the proposer not paying attention, probably going to leave that out. Multiple concurrent proposals for minimum amounts can be open at a time, so theres no telling if a proposal is created for the same amount, while another one is about to be approved. The dup proposal amount, _could_ pass after another in progress changes the amount and be valid after all.